### PR TITLE
Devel transport

### PR DESCRIPTION
**Pull request by PrzemekWirkus**
Changes related to [this devel_transport changes](https://github.com/ARMmbed/mbed-drivers/pull/162).

### CODE REVIEW

**Comment by PrzemekWirkus** [comment]
@bogdanm @bremoran @0xc0170 Please review!

**Comment by bogdanm** [comment]
Do we absolutely need to remove the old `test_env.h`? This will require a new major version on `mbed-drivers`, since I don't believe there's a way we can rewrite the old `test_env` in terms of the new one (they're not at all compatible, right?)

**Comment by 0xc0170** [comment]
@PrzemekWirkus please add details of source/test_env.cpp removal. I got the same question as @bogdanm above

**Comment by PrzemekWirkus** [comment]
@bogdanm @0xc0170 Guys, just like Bogdan said we will keep ```test_env``` in ```v1.x.y```.
I will make sure ```greentea-client/test_env.*``` doesn't produce linking errors when used together! :)

**Comment by 0xc0170** [comment]
> @bogdanm @0xc0170 Guys, just like Bogdan said we will keep test_env in v1.x.y.
I will make sure greentea-client/test_env.* doesn't produce linking errors when used together with mbed-drivers/test_env.* :)

:clap:

File: module.json [hunk 1]
```diff
@@ -35,7 +35,12 @@
     "ualloc": "^1.0.0",
     "minar": "^1.0.0",
     "core-util": "^1.0.0",
-    "compiler-polyfill": "^1.1.0"
+    "compiler-polyfill": "^1.1.0",
+    "greentea-client": "~0.1.4"
+  },
+  "testDependencies": {
+    "unity": "^2.0.0",
+    "greentea-client": "~0.1.4"
```
Comment (bogdanm) [inline module.json]: `greentea-client` is already listed as an unconditional dependency, so it doesn't need to be listed as a test dependency too.

File: source/retarget.cpp [hunk 2]
```diff
@@ -266,7 +291,14 @@ extern "C" int PREFIX(_write)(FILEHANDLE fh, const unsigned char *buffer, unsign
 #ifdef YOTTA_CFG_DEBUG_OPTIONS_COVERAGE
     else if(coverage_report && fh == gcov_fd) {
         for (unsigned int i = 0; i < length; i++) {
-            printf("%02x", buffer[i]);
+            if (0x00 == buffer[i]) {
+                // Very simple coverage stream compression
+                // Observation: About 30-35% of coverage stream are bytes with value 0x00
+                // Greentea will pick up '.' and replace with "00"
```
Comment (bogdanm) [inline source/retarget.cpp]: I believe this comment is backwards ?

**Comment by bogdanm** [comment]
OK, this is ready to be merged now. Some comments for people who don't know the whole story: the changes from this PR are meant to take in the new `greentea-client` repository, but also to keep the previous `mbed-drivers/test_env.h` and `source/test_env.cpp` to ensure compatibility with previous code. Some tests appear to have been removed, but actually they have been refactored as unity test cases. Once this PR is merged, I intend to release a new **minor** version of `mbed-drivers`.
I've added my comments to this PR; please help in reviewing this and merging it. Cc @bremoran @0xc0170 @adbridge

**Comment by bogdanm** [comment]
I would also add a `#warning` in `test_env.h` stating that this header is obsolete and `greentea-client/test_env.h` should be used instead.

**Comment by bogdanm** [comment]
Bump for review.
@0xc0170 @bremoran @mjs-arm @adbridge

**Comment by 0xc0170** [comment]
> I would also add a #warning in test_env.h stating that this header is obsolete and greentea-client/test_env.h should be used instead.

+1

Side question - @PrzemekWirkus When mbedgt is published ? I am running the latest version from the pypi, and of course this fails : 
```
+---------------+---------------+---------------------------------+--------+--------------------+-------------+
| target        | platform_name | test                            | result | elapsed_time (sec) | copy_method |
+---------------+---------------+---------------------------------+--------+--------------------+-------------+
| frdm-k64f-gcc | K64F          | mbed-drivers-test-c_strings     | ERROR  | 4.28               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-dev_null      | ERROR  | 4.27               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-echo          | ERROR  | 4.31               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-generic_tests | ERROR  | 4.28               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-rtc           | ERROR  | 4.29               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-stl_features  | ERROR  | 4.27               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-ticker        | ERROR  | 4.28               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-ticker_2      | ERROR  | 4.29               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-ticker_3      | ERROR  | 4.31               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-timeout       | ERROR  | 4.29               | shell       |
| frdm-k64f-gcc | K64F          | mbed-drivers-test-wait_us       | ERROR  | 4.28               | shell       |
+---------------+---------------+---------------------------------+--------+--------------------+-------------+
```
I linked this branch to one repo which was ported by your team to new async mbedgt, timeouted all tests. I got one repo still using old tests - tests are ok with this branch linked in, seems like no regression with this changes :+1:

File: CHANGELOG.md [hunk 3]
```diff
@@ -3,6 +3,25 @@ All notable changes to this project will be documented in this file.
 This project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased]
+### Added
+- Remove source/test_env.cpp and deprecate mbed-drivers/test_env.h in this module.
```
Comment (0xc0170) [inline CHANGELOG.md]: > Remove source/test_env.cpp

This is not true anymore? we got test_env.cpp here

Comment (PrzemekWirkus) [inline CHANGELOG.md]: Yes, please change it.

**Comment by PrzemekWirkus** [comment]
> Side question - @PrzemekWirkus When mbedgt is going to be published ? I am running the latest version from the pypi (0.1.19), and of course this fails : 

@0xc0170 Martin, new tools will be published when you publish new release.
You can use greentea and htrun from master (label v0.2.0).

**Comment by PrzemekWirkus** [comment]
@bogdanm Added warning stuff here: https://github.com/ARMmbed/mbed-drivers/pull/175

**Comment by bogdanm** [comment]
I would've liked to have the `#warning` as part of this PR. Also, the other comments are not resolved. I'm going to update this PR with a new commit fixing these issues in the next 15 mins or so.

**Comment by bogdanm** [comment]
https://github.com/ARMmbed/mbed-drivers/commit/dcbd4259c73b13840d45fd259d813f4d152e7864 removes the duplicated dependency on greentea-client, adds a warning to test_env.h and updates the CHANGELOG. This is ready to be merged as far as i'm concerned. Waiting for a final OK from either @bremoran or @0xc0170

**Comment by 0xc0170** [comment]
+1

**Comment by bogdanm** [comment]
Published version 1.2.0

**Pull request merged**